### ISSUE TITLE
FR: Backup FROM and ftp or SSH connected server

### ISSUE THREAD

**Issue by Germar** [bit-team/backintime#213]
It would be really useful to be able to backup FROM a server (webserver) down to my local desktop PC.

Doing an incremental backup over ftp or SSH would be the perfect solution for a lot of webdesigners. We need a reliable backup where we can go back in time in case of a break-in on the server, a harddisk crash, a bug in the system that cause failures etc.

It is not enough to do a simple sync that would overrwrite the local copy. This would be disaster if the original got corrupted as you would end up with a corrupted backup.

Also, it is important that it is possible to only backup the changed files as a full backup would use too much time and too much bandwith.

Just wish it would also be able to do the same with the MySQL database :-) But I guess that is a totally different talk...

**Comment by Germar** [comment]
(by danleweb)
You can use sshfs/curlftpfs to mount your server and then use BIT to backup mounted point.

**Comment by Germar** [comment]
(by danleweb)
You can use sshfs/curlftpfs to mount your server and then use BIT to backup mounted point.

**Comment by Germar** [comment]
(by oceanwatcher)
Yes... If it is supported at the remote end.

So I have to do this for all my 23 domains...

Would it not be easier to build this into the software so the software opens a session to the remote server, do the job and shut it down?

**Comment by Germar** [comment]
(by danleweb)
You can user user.callback to mount/unmount your server. This is for backup process.
But for the GUI you have to do it manually, at least for the moment.

**Comment by Germar** [comment]
(by ejacquot)
Hi all,

I think using sshfs to Back-In-Time is a bad and wrong idea :

First, there are some systems doesn't support FUSE ans SSHFS.
Second, Backintime seems to be an embedded rsync frontend. What about rsync over ssh ?

No time to paste any diff and don't use the trunk branch, but it will be usefull to do. So tell me your minds ;)

**Comment by Germar** [comment]
(by 0cs935kb517wwmwa7m9428daadkye-m9u2-wz6bkyhu4uqpfausw0ege9b0y33eg)
I hope this function will came next time...

**Comment by juliencarnot** [comment]
Hi,
I've been trying to backup some folders on my server to my computer using @danleweb's method of mounting them with SSHFS, but this results in an empty backup.

Folders are mounted and accessible to my user, and I can browse their content in BIT's gui, so I'm assuming they are accessible. My backup profile uses the same user, but the backup log shows no error and no sign of exploring the folders content. Not a single file is backed-up. Is this method confirmed to work, and if so, do you have an idea of why it wouldn't work for me? Thanks!

**Comment by Germar** [comment]
`sshfs` does not support hard-links. Not sure why it was empty but anyways, it would not be incremental. Please try BiT's SSH-Mode. It will also use `sshfs` internal. But only for browsing files. New snapshots are done by running `rsync` on the remote host.

Take a look at `man backintime` section SSH on how to set up. Or use current master. This will/should (fingers crossed) set up keys automatically.

**Comment by juliencarnot** [comment]
Hi Germar, thanks for your quick answer!

Sorry I haven't made my use case clear enough: I want to backup folders **from** my server **to** my computer (to have extra offline copies...). I've read the man and looked at the settings GUI, but I can't find a way to set a distant/SSH source and a local destination in SSH-Mode, did I miss something?

**Comment by Germar** [comment]
Oh, sorry. I didn't notice the context.

No, you didn't miss it. There is no such function for remote source (yet).

I just double checked with a mounted remote source and it worked fine here. Please open a new issue and post the output of `backintime --debug --profile-id X backup` in there.

**Comment by juliencarnot** [comment]
My bad: the folder I was trying to backup was mounted to a path belonging to a symlink, and not to the target folder, so the source folder I had defined was not seeing the sshfsed content, resulting in the empty backup.

**Comment by jaknz** [comment]
Hi, I found this thread and figured I would revive it instead of opening a new issue. Is there any change in this, or is using a remote SSH source still a wishlist item?

**Comment by emtiu** [comment]
Hi @jaknz, thanks for the report and finding this Issue. It's still not implemented, and development has just restarted after a long pause. So this will probably remain a request for the time being :)

**Comment by protist** [comment]
I would also like to backup from a remote server. I'll call this the _source_, and it will back up to the _target_. In my case, I would prefer my credentials for the _target_ not to be shared with the _source_. Hence, I'd prefer to initiate backintime from the _target_ server instead (i.e. backup a remote _source_). Here is how I got it working.

Firstly, set up ssh access to the _source_:

1. On the _target_, install sshfs
2. Set up ssh from the _target_ into the _source_ using a key with no passphrase. In my case, I generated the key on the _target_ using `ssh-keygen -t rsa -b 4096`, then saved it to `/root/.ssh/root@source_server2_rsa`
3. Copy the public key from _target_'s `/root/.ssh/root@source_server_rsa.pub` to the _source_  `/root/.ssh/authorized_keys`, or use `ssh-copy-id` if you prefer. (I want to back up everything, hence root on the _source_. I also want to be able to restore directly from the rsync backup, hence root on the _target_.)
4. Check you can ssh from the _target_ okay, with `sudo ssh root@192.168.1.92 -i /root/.ssh/root@source_server2_rsa`
5. Make a path to mount the _source_ later, e.g. `mkdir /root/mnt`

Set up backintime on the _target_:
1. Edit the config, e.g. at `/root/.config/backintime/config`. Some important lines are listed here
```
profile2.name=source_server2
profile2.snapshots.include.1.value=/root/mnt/source_server2
profile2.snapshots.path.host=source_server2
```
2. Also note that the `exclude` lines will now need to be relative paths, since you are mounting the _source_ in a subdirectory. e.g. if you are backing up all of the _source_, i.e. `/`, then instead of
```
profile2.snapshots.exclude.1.value=/media/*
```
you would want
```
profile2.snapshots.exclude.1.value=media/*
```
3. create a [user-callback](https://github.com/bit-team/user-callback) e.g. at `/root/.config/backintime/user-callback`, and make it executable. This will `sshfs` mount the _source_. FWIW I manually `mkdir` and `rmdir` the mount location, just in case there's some error, and the connection is interrupted. This is probably overkill, but I can't think of a drawback. My file contains the following.
```
#!/bin/sh
# exit on error
set -e

profile_id="$1"
profile_name="$2"
reason="$3"

# Manually map the IPs for each profile
case "$profile_name" in
source_server2)
  address="192.168.1.92"
  ;;
source_server3)
  address="192.168.1.93"
  ;;
esac

# Location to mount the source
source="/root/mnt/${profile_name}"

# Do the mount
case $reason in
1) # Backup process begins
  mkdir "${source}" # just in case there are mounts still left from a previous error
  sshfs -o no_contain_symlinks "root@${address}:/" "${source}" -o "IdentityFile=/root/.ssh/root@${profile_name}_rsa"
  ;;
2 | 4) # Backup process ends OR Error
  set +e # don't exit on error. Even if you can't umount, still rmdir
  umount "${source}"
  rmdir "${source}"
  ;;
esac
```

This is working well for me. The only problem that I've come across so far is that AFAICT you can't seem to do a user-callback with `backintime restore`. So presumably you'd have to manually `sshfs` mount the _source_ before restoring files.

**EDIT**
Another issue I've found is that the snapshots are stored at `backintime/<HOST>/<USER>/<PROFILE_ID>/<DATE>/backup`, but then within the subdirectory defined by the mount point. i.e. for me the root of the _source_ is stored at `backintime/source_server2/root/2/20240618-173228-313/backup/root/mnt/source_server2/`. This is a bit annoying, because it hardcodes the mount point into the snapshot location. Ideally there would be a way to not have the mount point in the snapshot path, i.e. store the root directly at `backintime/source_server2/root/2/20240618-173228-313/backup/`. However I can't seem to find a way to do this.

**EDIT2**
It looks like the user-callback doesn't actually run all the time there is an error. Supposedly it should run with `$3=4` when there is an error. If the mount directory already exists, and you try to backup, the script will fail as expected at the `mkdir` line. Then the backup will not proceed, as expected. However, given that backintime has encountered an error, I would expect the user-callback to run again, with `$3=4`. This would then `rmdir`, so that the backup can work the next time. However, the user-callback is _not_ called on this error unfortunately.

**EDIT3**
Added `-o no_contain_symlinks`, otherwise this breaks with sshfs 3.7.6

**Comment by emtiu** [comment]
Thank you @protist for the detailed report! It's the kind of solution I would have recommended, to mount what you want to access remotely.

Your complaint about the target path is related (but not identical) to #1304. I don't think we have an open Issue for that yet, you might want to open one for the "Wishlist".

Also, I don't think there's an Issue about the user-callback behavior you encountered (and I'm not 100% certain it's a bug), so opening an Issue for that might also help.

**Comment by protist** [comment]
Thanks @emtiu and done!

**Comment by buhtz** [comment]
Hello Protist,

> I would also like to backup from a remote server. I'll call this the source, and it will back up to the target. In my case, I would prefer my credentials for the target not to be shared with the source. Hence, I'd prefer to initiate backintime from the target server instead (i.e. backup a remote source). Here is how I got it working.

I don't understand this use case, especially the first sentence. Can you explain it in a more plain language please. How many machines? On which machine is BIT running?

If you want to backup data from a remote server you have to mount its folders (e.g. via SSH, or something else) to the machine where BIT is running.

It seems I do miss a point?

**Comment by protist** [comment]
Hi @buhtz, and thanks for the questions. I have two machines.

1. *Source* server runs a custom distro running potentially insecure software.
2. *Target* is my primary server. It's hardened, etc., and runs BIT.
3. I would like to back up *source* to a remote server (i.e. to *target*), but I don't want *source* to be able to access *target* in any meaningful way. Since *source*'s security is questionable, if it were to be compromised, this would potentially compromise my main server *target*.
4. Instead, I want it the other way around, i.e. *target* to access *source*. *Target* is unlikely to be compromised, and even if it is, I don't really care if *source* is subsequently compromised.
5. Hence, I'd like BIT on *target* to backup the files on *source*, e.g. by connecting via ssh.

As per my other post, currently *target* mounts *source* using `sshfs` and the scripts I've posted. Ideally BIT would be able to do this automatically instead.

**Comment by buhtz** [comment]
Thank you, now I understand better.

> As per my other post, currently _target_ mounts _source_ using `sshfs` and the scripts I've posted. Ideally BIT would be able to do this automatically instead.

I am assuming I still miss a point. Because as I understand the scenario you subscribing, it could easyilly be solved by just configuring a "SSH" backup profile.

So what hinders you to create such an SSH profile? BIT (running on "target") is able to access "source" via SSH.

**Comment by protist** [comment]
No worries. Thanks for chasing this up @buhtz .

> So what hinders you to create such an SSH profile? BIT (running on "target") is able to access "source" via SSH.

I'm not sure I understand now! Doesn't BIT only allow you to SSH *to* the backup storage? It won't let you SSH data *from* the source that you want to backup? i.e. in my situation, let's assume I only want the secure server *target* to have the SSH credentials of *source*. Then this means I can only save BIT backups remotely, i.e. backup *target* → *source*. Whereas I want to backup *source* → *target*.

**Comment by buhtz** [comment]
> Doesn't BIT only allow you to SSH _to_ the backup storage?

Correct. Now I see my understanding problem. 🤣 

> It won't let you SSH data _from_ the source that you want to backup?

Correct. The backup source (the files and dirs to be backuped up) need to somehow accessible in the "local" file system. So before starting the backup job you have to mount the remote source (via SSH, NFS, Samba, SomeOtherFancyProtocol) to your file system.

This is what user-callback scripts are for. They don't work perfect currently but they do  work.

Would this be a solution?

---

Technically this issue seems to be about support remote source via several protocols (rsync-server #224, sshfs, and some more) and without pre-mounting the source. Am I right so far?

**Comment by protist** [comment]
> This is what user-callback scripts are for. They don't work perfect currently but they do work.
> Would this be a solution?

Yep, it might have gotten lost in this long thread, but that's what I explained how to do in my [first post](https://github.com/bit-team/backintime/issues/213#issuecomment-2176036154).

It's just quite fiddly, and probably prone to errors. I just think it would be nicer if it were integrated with better checks.

**Comment by buhtz** [comment]
Hello protist,

OK, thank you for being with me and your patience. Sometimes it takes some time until I get it. 😃 With my current understanding I would reject such a feature in BIT, even if someone else would contribute it.

Of course such a feature cost resources to being implemented and also to be maintained. There is also the question which protocols (SSH, RSYNC, NFS, ...) to support. And BIT's codebase currently is definitive not in the state to adapt to such feature. But that is not the primary reason.

Doing backups of several remote machines is an often seen feature in advanced/enterprise backup software. I see this as "another level" of backup software. BIT is not for enterprise users. It is for desktop end users. I accept, that some of these "end users" might have special needs, especially when the are in the GNU/Linux universe. But these might be rare.

In short: The feature does not fit to the principles (or concept?) of the project, regarding the targeted user group.

If there are no new arguments coming up I would close the feature request as not planed.

Regards,
Christian

**Comment by protist** [comment]
No worries at all @buhtz. FWIW I'm certainly not enterprise; I'm just running it on my home servers. I get your point about maintenance though. Totally happy with whatever you decide! I didn't open the bug, so was more commenting here to give details about the (slightly complicated and buggy) workaround.

**Comment by buhtz** [comment]
Closing this ticket based on the comment above.
Feel free to reopen if the problem still exists. Thank you for your
efforts. If you have any further questions, ideas or encounter any
other issues, please don't hesitate to let us know.

Best regards,
Christian
